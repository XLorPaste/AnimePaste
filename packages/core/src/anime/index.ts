export * from './anime';

export * from './types';
