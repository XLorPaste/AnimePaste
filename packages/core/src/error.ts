export class AnimeSystemError extends Error {
  public readonly detail: string;

  constructor(detail: string) {
    super(detail);
    this.detail = detail;
  }
}
