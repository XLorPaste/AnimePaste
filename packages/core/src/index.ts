export * from './space';

export * from './anime';

export * from './plugin';

export * from './system';

export * from './error';

export * from './utils';
