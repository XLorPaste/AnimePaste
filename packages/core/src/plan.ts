export interface Plan {}

export async function loadPlan(plans: string[]) {
  return [] as Plan[];
}
