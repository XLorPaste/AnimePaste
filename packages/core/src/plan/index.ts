export * from './plan';

export * from './types';
