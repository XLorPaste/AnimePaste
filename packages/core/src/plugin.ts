import type { Breadc } from 'breadc';
import type { AnyZodObject } from 'zod';

import type { Anime, LocalFile, LocalVideo } from './anime';
import type { AnimeSpace, Plan, PluginEntry } from './space/schema';
import type { AnimeSystem, IntrospectOptions, RefreshOptions } from './system';

type MayPromise<T> = T | Promise<T>;

export interface Plugin {
  /**
   * The name of your plugin
   */
  name: string;

  /**
   * Options to create this plugin
   */
  options: PluginEntry;

  prepare?: {
    /**
     * Prepare anime space configurations
     */
    space?: (space: AnimeSpace) => MayPromise<void>;

    /**
     * Prepare anime space plans
     */
    plans?: (space: AnimeSpace, plans: Plan[]) => MayPromise<void>;
  };

  schema?: {
    space?: AnyZodObject;

    plan?: AnyZodObject;
  };

  /**
   * Extend command line interface
   */
  command?: (system: AnimeSystem, cli: Breadc<{}>) => MayPromise<void>;

  introspect?: {
    pre?: (system: AnimeSystem, options: IntrospectOptions) => MayPromise<void>;

    handleUnknownFile?: (
      system: AnimeSystem,
      anime: Anime,
      file: LocalFile,
      options: IntrospectOptions
    ) => MayPromise<LocalVideo | undefined>;

    handleUnknownVideo?: (
      system: AnimeSystem,
      anime: Anime,
      video: LocalVideo,
      options: IntrospectOptions
    ) => MayPromise<LocalVideo | undefined>;

    post?: (
      system: AnimeSystem,
      options: IntrospectOptions
    ) => MayPromise<void>;
  };

  refresh?: {
    pre?: (system: AnimeSystem, options: RefreshOptions) => MayPromise<void>;

    refresh?: (
      system: AnimeSystem,
      anime: Anime,
      options: RefreshOptions
    ) => MayPromise<void>;

    post?: (system: AnimeSystem, options: RefreshOptions) => MayPromise<void>;
  };

  writeLibrary?: {
    pre?: (system: AnimeSystem, anime: Anime) => MayPromise<void>;

    post?: (system: AnimeSystem, anime: Anime) => MayPromise<void>;
  };
}
