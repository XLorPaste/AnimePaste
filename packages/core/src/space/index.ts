export * from './schema';

export * from './plan';

export * from './space';
