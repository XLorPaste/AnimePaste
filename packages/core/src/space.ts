import fs from 'node:fs';
import path from 'node:path';

import { parse, stringify } from 'yaml';

import { AnimeSystemError } from './error';

const configFilename = `./anime.yaml`;

const DefaultStorageDirectory = `./anime`;

export interface AnimeSpace {
  readonly root: string;

  readonly storage: string;

  readonly preference: Preference;

  readonly plans: string[];

  readonly plugins: PluginEntry[];
}

export interface Preference {
  keywords: {
    order: Record<string, string[]>;
    exclude: string[];
  };
}

export interface PluginEntry {
  name: string;

  options: Record<string, any>;
}

export async function loadSpace(_root: string) {
  const root = path.resolve(_root);

  const configPath = path.join(root, configFilename);
  if (fs.existsSync(root) && fs.existsSync(configPath)) {
    // Load space directory
    const configContent = fs.readFileSync(configPath, 'utf-8');
    const config = parse(configContent);

    const storageDirectory: string = config.storage ?? DefaultStorageDirectory;
    const plans = (config.plans ?? []).map((p: string) => path.join(root, p));
    const space: AnimeSpace = {
      root,
      storage: path.resolve(root, storageDirectory),
      preference: config.preference ?? {},
      plans,
      plugins: ((config.plugins ?? []) as any[]).map((p) => {
        if (typeof p === 'string') {
          return { name: p, options: {} };
        } else {
          const o = { ...p };
          const name = o.name;
          delete o.name;
          return { name, options: { ...o } };
        }
      })
    };

    // Validate space directory
    await validateSpace(space);
    return space;
  } else {
    // Create new space directory
    return makeNewSpace(root);
  }
}

async function validateSpace(space: AnimeSpace) {
  try {
    fs.accessSync(space.root, fs.constants.R_OK | fs.constants.W_OK);
  } catch {
    throw new AnimeSystemError(`Can not access AnimePaste space directory`);
  }
  try {
    fs.accessSync(space.storage, fs.constants.R_OK | fs.constants.W_OK);
  } catch {
    throw new AnimeSystemError(`Can not access local anime storage directory`);
  }
  for (const p of space.plugins) {
    if (!p.name || !p.options || typeof p.options !== 'object') {
      throw new AnimeSystemError(`Plugin configuration may be incorrect`);
    }
  }
  for (const p of space.plans) {
    if (!fs.existsSync(p)) {
      throw new AnimeSystemError(`Plan ${p} config is not found`);
    }
  }
  return true;
}

async function makeNewSpace(root: string) {
  const space: AnimeSpace = {
    root,
    storage: path.join(root, DefaultStorageDirectory),
    preference: {
      keywords: {
        order: {
          format: ['mp4', 'mkv'],
          language: ['简', '繁'],
          resolution: ['1080', '720']
        },
        exclude: ['HEVC']
      }
    },
    plans: [],
    plugins: [
      { name: 'animegarden', options: {} },
      { name: 'download', options: { directory: './download' } }
    ]
  };

  await fs.promises.mkdir(space.root, { recursive: true }).catch(() => {});
  await fs.promises.mkdir(space.storage, { recursive: true }).catch(() => {});
  await fs.promises
    .mkdir(path.join(space.root, './plans'), { recursive: true })
    .catch(() => {});

  await fs.promises.writeFile(
    path.join(space.root, configFilename),
    stringify({
      ...space,
      root: undefined,
      storage: DefaultStorageDirectory,
      plugins: [
        { name: 'animegarden' } as PluginEntry,
        { name: 'download', directory: './download' } as unknown as PluginEntry
      ]
    }),
    'utf-8'
  );

  return space;
}
