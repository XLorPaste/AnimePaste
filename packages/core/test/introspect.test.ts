import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { describe, it, expect } from 'vitest';

import { createAnimeSystem, loadSpace } from '../src';

const __dirname = path.join(fileURLToPath(import.meta.url), '../');

describe('Introspect', () => {
  it('should work', async () => {
    const root = path.join(__dirname, './fixtures/space');
    const space = await loadSpace(root, async (entry) => {
      return undefined;
    });
    const system = await createAnimeSystem(space);
    const animes = await system.introspect();
    expect(await Promise.all(animes.map((a) => a.library()))).toMatchSnapshot();
    expect(await Promise.all(animes.map((a) => a.list()))).toMatchSnapshot();
  });
});
