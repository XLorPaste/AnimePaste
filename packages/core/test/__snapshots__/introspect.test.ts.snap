// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Introspect > should work 1`] = `
[
  Anime {
    "_lib": {
      "bgmId": "323651",
      "title": "熊熊勇闯异世界 Punch!",
      "videos": [],
    },
    "directory": "/Users/xlor/Repos/AnimePaste/packages/core/test/fixtures/space/anime/熊熊勇闯异世界 Punch!",
    "dirty": false,
    "plan": {
      "bgmId": "323651",
      "date": 2023-04-01T13:00:00.000Z,
      "fansub": [
        "Lilith-Raws",
        "ANi",
      ],
      "keywords": {
        "exclude": [],
        "include": [
          [
            "熊熊勇闯异世界 Punch!",
          ],
        ],
      },
      "season": 2,
      "state": "onair",
      "title": "熊熊勇闯异世界 Punch!",
      "type": "番剧",
    },
  },
  Anime {
    "_lib": {
      "bgmId": "404804",
      "title": "天国大魔境",
      "videos": [],
    },
    "directory": "/Users/xlor/Repos/AnimePaste/packages/core/test/fixtures/space/anime/天国大魔境",
    "dirty": false,
    "plan": {
      "bgmId": "404804",
      "date": 2023-04-01T13:00:00.000Z,
      "fansub": "SweetSub",
      "keywords": {
        "exclude": [],
        "include": [
          [
            "天国大魔境",
          ],
        ],
      },
      "season": 1,
      "state": "onair",
      "title": "天国大魔境",
      "type": "番剧",
    },
  },
  Anime {
    "_lib": {
      "bgmId": "376703",
      "title": "偶像大师 灰姑娘女孩 U149",
      "videos": [],
    },
    "directory": "/Users/xlor/Repos/AnimePaste/packages/core/test/fixtures/space/anime/偶像大师 灰姑娘女孩 U149",
    "dirty": false,
    "plan": {
      "bgmId": "376703",
      "date": 2023-04-01T13:00:00.000Z,
      "fansub": "喵萌奶茶屋",
      "keywords": {
        "exclude": [
          "闪耀色彩",
        ],
        "include": [
          [
            "偶像大师",
            "iDOLM@STER",
          ],
          [
            "灰姑娘女孩",
          ],
          [
            "U149",
          ],
        ],
      },
      "season": 1,
      "state": "onair",
      "title": "偶像大师 灰姑娘女孩 U149",
      "type": "番剧",
    },
  },
]
`;
